porting stm32f2

qianfan-Zhao opened the pull request:
Tested with my custom board(stm32f207ve):

- [x] blinky (gpio)
- [x] hello_world (serial)

qianfan-Zhao committed:
hal: stm32f2x: Add HAL for the STM32F2x series

This adds the HAL code needed for the SoC for the stm32f2 series.

Origin: ST Microelectronics
License: BSD-3-Clause
URL: http://www.st.com/en/embedded-software/stm32cubef2.html
Commit: 1.7.0
Purpose: HAL layer for stm32f2
Maintained-by: External

Signed-off-by: qianfan Zhao <qianfanguijin@163.com>

qianfan-Zhao committed:
arm: stm32f2: Add support for stm32f2 series

Add necessary kconfig and DTS files for stm32f2 series

Signed-off-by: qianfan Zhao <qianfanguijin@163.com>

qianfan-Zhao committed:
drivers: stm32-gpio: Provide GPIO driver for stm32f2

Add necessary GPIO defines for the stm32f2 series soc

Signed-off-by: qianfan Zhao <qianfanguijin@163.com>

qianfan-Zhao committed:
driver: pinmux: Add pinmux driver for stm32f2

Add driver header file for the stm32f2 pinmux

Signed-off-by: qianfan Zhao <qianfanguijin@163.com>

qianfan-Zhao committed:
drivers: clock_control: Add support for stm32f2

Add clock control support for the stm32f2

Signed-off-by: qianfan Zhao <qianfanguijin@163.com>

qianfan-Zhao committed:
driver: interrupt_controller: Add support for stm32f2

Add kconfig and c code for stm32f2 interrupt controller driver

Signed-off-by: qianfan Zhao <qianfanguijin@163.com>

erwango commented on drivers/clock_control/stm32_ll_clock.c:
```diff
@@ -56,7 +56,8 @@ static inline int stm32_clock_control_on(struct device *dev,
 		LL_AHB1_GRP1_EnableClock(pclken->enr);
 		break;
-#if defined(CONFIG_SOC_SERIES_STM32L4X) || defined(CONFIG_SOC_SERIES_STM32F4X)
+#if defined(CONFIG_SOC_SERIES_STM32L4X) || defined(CONFIG_SOC_SERIES_STM32F4X) \\
```
Use '\' instead of '\\'.
This is making the CI failing

zhaozhongchen replied:
yes '\\' is my mistake.

this pr was still working and can't build. I will fix those later. But I don't have a wellknow board so I haven't add a board.

zhaozhongchen replied:
yes it is.

erwango replied:
Well, this is an issue for upstream, since we want to be able to build and test upstream code. Is there a way you could get a well know board?

zhaozhongchen replied:
I think I can buy one.

erwango reviewed (changes requested):
Had a first look and looks nice.
Please fix CI, then we'll review in detail.

One thing to correct is that you should focus your changes on SoC variant that are supported by the board you introduce, so this can be tested. We don't want to introduce code that could not be tested be cause we miss board to test.

qianfan-Zhao committed:
driver: uart_stm32: add support for stm32f2 series

clear RXNE flag in fifo_read, remove TEACK and REACK
check when uart_stm32_init because stm32f2 doesn't
has those flags.

Signed-off-by: qianfan Zhao <qianfanguijin@163.com>

erwango commented on ext/hal/st/stm32cube/stm32f2xx/drivers/include/stm32f2xx_hal_conf_template.h:
```diff
@@ -0,0 +1,401 @@
+/**
```
Please remove this file

erwango commented on arch/arm/soc/st_stm32/stm32f2/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,36 @@
+
+endif # GPIO_STM32
+
+if USB
```
Please remove USB from this commit

erwango commented on arch/arm/soc/st_stm32/stm32f2/Kconfig.defconfig.stm32f205xx:
```diff
@@ -0,0 +1,18 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+if SOC_STM32F205XB || SOC_STM32F205XC || SOC_STM32F205XE || SOC_STM32F205XF || SOC_STM32F205XG
```
Please focus on the SoC you're introducing.

zhaozhongchen replied:
sorry I don't know waht is foucus, can you explain more clearer? seems that if I use nucleo-f207, I need remove  all soc except f207?

erwango replied:
I meant "limit to", so yes, if you target nucleo_f207ZG, please keep only f207xx

erwango commented on arch/arm/soc/st_stm32/stm32f2/Kconfig.soc:
```diff
@@ -0,0 +1,42 @@
+prompt "STM32F2X MCU Selection"
+depends on SOC_SERIES_STM32F2X
+
+config SOC_STM32F205XB
```
Focus on the introduced SoC

erwango commented on arch/arm/soc/st_stm32/stm32f2/dts.fixup:
```diff
@@ -0,0 +1,25 @@
+
+#define CONFIG_NUM_IRQ_PRIO_BITS	        ARM_V7M_NVIC_E000E100_ARM_NUM_IRQ_PRIORITY_BITS
+
+#define CONFIG_UART_STM32_PORT_1_BASE_ADDRESS	ST_STM32_USART_40011000_BASE_ADDRESS
```
Move UART related to the commit introducing UART driver

erwango commented on arch/arm/soc/st_stm32/stm32f2/dts.fixup:
```diff
@@ -0,0 +1,25 @@
+#define CONFIG_UART_STM32_PORT_2_NAME		ST_STM32_USART_40004400_LABEL
+#define PORT_2_IRQ				ST_STM32_USART_40004400_IRQ_0
+
+#define CONFIG_USB_BASE_ADDRESS			ST_STM32_OTGFS_50000000_BASE_ADDRESS
```
Move USB related to the commit introducing USB driver

erwango commented on arch/arm/soc/st_stm32/stm32f2/gpio_registers.h:
```diff
@@ -0,0 +1,52 @@
+ * Chapter 7: System configuration controller (SYSCFG)
+ */
+
+struct stm32f2x_gpio {
```
Move these to GPIO introduction commit

erwango commented on arch/arm/soc/st_stm32/stm32f2/soc_gpio.c:
```diff
@@ -0,0 +1,144 @@
+/*
```
Move to GPIO

erwango reviewed (changes requested):
I had a first look.

Tow main comments in order to minimize the introduced code and make the review faster
* Focus on the SoC you're currently introducing
* Split the commits based on the components you're introducing (hal, soc, clock, gpio, uart, ...)
* Do not introduce code that will not be tested yet (USB, I2C, ..)

Cf #7284 in case you have any doubts

qianfan-Zhao committed:
boards: arm: add st nucleo-f207zg

tested example: blinky, hello_world

Signed-off-by: qianfan Zhao <qianfanguijin@163.com>

zhaozhongchen commented:
@erwango I don't have nucleo-f207zg board and tested this pr on my custom boards with examples: blinky, hello_world, cdc_acm. 

And my custom board is f207ve, so I leave STM32F207XE and STM32F207XZ soc on this pr, and delete others.

erwango commented:
@zhaozhongchen : Ok, I've found one nucleo_f207zg board, I'll do the tests when I have some time (not before next week)

> And my custom board is f207ve, so I leave STM32F207XE and STM32F207XZ soc on this pr, and delete others.

Can you clarify why you need to keep STM32F207XZ?

zhaozhongchen commented:
> Can you clarify why you need to keep STM32F207XZ?

@erwango sorry it's my mistake, I leave STM32F207XE(my custom board) and STM32F207XG(nucleo-f207zg) in the code, not STM32F207XZ.

ibisz commented:
Can you keep in this PR the STM32F215XG too?
I have the possibility do the tests with a STM32F215ZG based custom board.

zhaozhongchen commented:
@ibisz Thanks for you that you can test this pr. I think we'd better don't leave STM32F215 soc on this pr, because there doesn't has any board in 'boards/' dir that is based on STM32F215 SOC. 

But I pushed [stm32f215](https://github.com/zhaozhongchen/zephyr/tree/stm32f215zg) branch on my local repository, and I will keep that branch updated with this pr.  You can tested using that branch.

Thanks

ibisz commented:
@zhaozhongchen: Thanks for the additional effort.
In any way, the project should support the custom boards handling in a more flexible way, taking into consideration the final goal to propagate it to the 'real life' hardware solutions.
I will start to use an extra-tree 'BOARD_ROOT' for the custom board definitions.

erwango commented:
> In any way, the project should support the custom boards handling in a more flexible way, taking into consideration the final goal to propagate it to the 'real life' hardware solutions.

@ibisz , this is right. Though, we should at least be able to track introduced code for which there is no board available in upstream for test. This require to have dedicated PR with information mentioning that code has been tested.

@zhaozhongchen , can you keep STM32F207XE introduction in a separate commit, and indicate it has been tested with your custom board?

zhaozhongchen commented:
I had updated this PR and now net example(http_server) tested passed on my custom board.

@erwango F207XE and F207XG is quite similer except SRAM and FLASH size. Is it ok that make this two soc together?

erwango commented:
@zhaozhongchen , I prefer to add it in a separate commit, with a message that it has been tested on custom board. We generally don't accept SoC w/o board in tree for test. Having them makes sense, but to do that, we need at least to be able to track clearly how it was introduced.

ibisz commented on arch/arm/soc/st_stm32/stm32f2/dts.fixup:
```diff
@@ -0,0 +1,49 @@
+#define CONFIG_UART_STM32_PORT_4_BAUD_RATE	ST_STM32_USART_40004C00_CURRENT_SPEED
+#define CONFIG_UART_STM32_PORT_4_IRQ_PRI	ST_STM32_USART_40004C00_IRQ_0_PRIORITY
+#define PORT_4_IRQ				ST_STM32_USART_40004C00_IRQ_0
+
```
Switching the console to UART4 generates:
zephyr/include/generated/generated_dts_board.h:185:47: error: 'ST_STM32_USART_40004C00_BASE_ADDRESS' undeclared here (not in a function); did you mean 'ST_STM32_UART_40004C00_BASE_ADDRESS'?
 #define CONFIG_UART_STM32_PORT_4_BASE_ADDRESS ST_STM32_USART_40004C00_BASE_ADDRESS
 
Correcting USART to UART in the subsequent lines the error disapears:

#define CONFIG_UART_STM32_PORT_4_NAME		ST_STM32_UART_40004C00_LABEL
#define CONFIG_UART_STM32_PORT_4_BASE_ADDRESS	ST_STM32_UART_40004C00_BASE_ADDRESS
#define CONFIG_UART_STM32_PORT_4_BAUD_RATE	ST_STM32_UART_40004C00_CURRENT_SPEED
#define CONFIG_UART_STM32_PORT_4_IRQ_PRI	ST_STM32_UART_40004C00_IRQ_0_PRIORITY
#define PORT_4_IRQ				ST_STM32_UART_40004C00_IRQ_0

zhaozhongchen replied:
Thank you, it's my mistake and I had fix it.

ibisz commented on arch/arm/soc/st_stm32/stm32f2/dts.fixup:
```diff
@@ -0,0 +1,49 @@
+#define CONFIG_UART_STM32_PORT_5_BAUD_RATE	ST_STM32_USART_40005000_CURRENT_SPEED
+#define CONFIG_UART_STM32_PORT_5_IRQ_PRI	ST_STM32_USART_40005000_IRQ_0_PRIORITY
+#define PORT_5_IRQ				ST_STM32_USART_40005000_IRQ_0
+
```
See previous comment:

#define CONFIG_UART_STM32_PORT_5_NAME		ST_STM32_UART_40005000_LABEL
#define CONFIG_UART_STM32_PORT_5_BASE_ADDRESS	ST_STM32_UART_40005000_BASE_ADDRESS
#define CONFIG_UART_STM32_PORT_5_BAUD_RATE	ST_STM32_UART_40005000_CURRENT_SPEED
#define CONFIG_UART_STM32_PORT_5_IRQ_PRI	ST_STM32_UART_40005000_IRQ_0_PRIORITY
#define PORT_5_IRQ				ST_STM32_UART_40005000_IRQ_0

qianfan-Zhao committed:
stm32f2: add stm32f207xg soc

Add stm32f207xg soc to make nucleo-f207zg work.

Signed-off-by: qianfan Zhao <qianfanguijin@163.com>

qianfan-Zhao committed:
stm32f2: add stm32f207xe soc

Used and tested(blinky, hello_world) on out of tree custom board.

Signed-off-by: qianfan Zhao <qianfanguijin@163.com>

zhaozhongchen commented:
@erwango please review again

erwango commented on arch/arm/soc/st_stm32/stm32f2/soc.h:
```diff
@@ -0,0 +1,47 @@
+
+#include "soc_irq.h"
+
+#ifdef CONFIG_SERIAL_HAS_DRIVER
```
Move to commit where you introduce serial driver

erwango commented on dts/arm/st/stm32f2-pinctrl.dtsi:
```diff
@@ -0,0 +1,50 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <dt-bindings/pinctrl/stm32-pinctrl.h>
```
Move to pinctrl commit

erwango commented on dts/arm/st/stm32f2.dtsi:
```diff
@@ -0,0 +1,272 @@
+	};
+
+	soc {
+		flash-controller@40023c00 {
```
Move to flash dedicated commit, with test indication

erwango commented on dts/arm/st/stm32f2.dtsi:
```diff
@@ -0,0 +1,272 @@
+			};
+		};
+
+		rcc: rcc@40023800 {
```
Either move to Clock_control driver introduction commit,
either do a dts specific introduction commit after the drivers introduction

erwango reviewed (commented):
@zhaozhongchen , PR looks good.
Some reordering/clean up before detailed review:
-device tree: Provide a dedicated commit after drivers introduction, add the .fixup at the same time. Only add nodes that you've been able to validate: remove spi, i2c, flash.
-Put the custom SoC at the end.
-Keep usb/ethernet out of this PR. You can add another PR, based on this one at the same time. Then for each driver you enable add a comment in the help section of the dedicated Kconfig flag (such as USB_DC_STM32 in usb case)

Reason for these requests is to simplify and fasten review.  And have a things clear in tree if anyone looks at lit later.
Thanks for this work !

ibisz commented:
@zhaozhongchen , Thanks for  the updates on stm32f215 branch on your local repository.
The PR actually is based on V1.11.0.
My advice is to rebase it together with this PR to the actual ‘master’ branch to keep it synchronized with the latest upstream enhancements.
After rebasing it, there are some changes required in 'dts.fixup', see below:
'PORT_1' to 'USART_1'
'PORT_2 to 'USART_2'
'PORT_3 to 'USART_3'
'PORT_6 to 'USART_6'
'PORT_4 to 'UART_4'
'PORT_5 to 'UART_5'
Time to time, to keep the PR up to date, this rebase is inevitable and consequently restarts the re-execution of the all, already completed, tests cases.

zhaozhongchen commented:
@ibisz Thanks! And I had update code to master and remove usb. eth commit according @erwango 's request.

ibisz commented:
@zhaozhongchen , Thanks for the updates on stm32f215 branch on your local repository.
The branch is tested with my custom board(stm32f215zg):
I think the branch is ready for a comit: 'stm32f2: add stm32f215zg soc'

A. Sample 'threads' (gpio, serial):
***** Booting Zephyr OS 1.12.0-rc1 *****
Toggle USR0 LED: Counter = 0
Toggle USR1 LED: Counter = 0
Toggle USR0 LED: Counter = 1
Toggle USR0 LED: Counter = 2
Toggle USR0 LED: Counter = 3
Toggle USR0 LED: Counter = 4
Toggle USR0 LED: Counter = 5
Toggle USR0 LED: Counter = 6
Toggle USR0 LED: Counter = 7
Toggle USR0 LED: Counter = 8
Toggle USR0 LED: Counter = 9
Toggle USR1 LED: Counter = 1
Toggle USR0 LED: Counter = 10
Toggle USR0 LED: Counter = 11

B. Sample  'hello_world' (serial):
***** Booting Zephyr OS 1.12.0-rc1 *****
Hello World! arm

zhaozhongchen commented:
@erwango got it and i like that way

dbkinder reviewed (approved):
LGTM, thanks!

erwango commented:
recheck

erwango commented on dts/arm/st/stm32f2.dtsi:
```diff
@@ -137,6 +137,60 @@
+			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00040000>;
+			interrupts = <39 0>;
+			status = "disabled";
+			label = "UART_3";
```
USART_3

erwango commented on dts/arm/st/stm32f2.dtsi:
```diff
@@ -137,6 +137,60 @@
+			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00020000>;
+			interrupts = <38 0>;
+			status = "disabled";
+			label = "UART_2";
```
USART_2

erwango commented on dts/arm/st/stm32f2.dtsi:
```diff
@@ -137,6 +137,60 @@
+			clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00004000>;
+			interrupts = <37 0>;
+			status = "disabled";
+			label = "UART_1";
```
USART_1

ibisz replied:
In the already committed ‘*.dtsi’ files for stm32fx the usartn and uartn are labelled with “UART_n”.
If this is a new requirement, I think that the usart6 should be labelled with “USART_6” too.

erwango replied:
This is right, my bad. Let's keep UART for now. Will remove these comments

erwango commented on boards/arm/nucleo_f207zg/doc/nucleo_f207zg.rst:
```diff
@@ -0,0 +1,180 @@
++-----------+------------+-------------------------------------+
+| GPIO      | on-chip    | gpio                                |
++-----------+------------+-------------------------------------+
+| I2C       | on-chip    | i2c                                 |
```
Set only what is supported on the board right now (remove I2C, USB, PWM)

erwango commented on boards/arm/nucleo_f207zg/nucleo_f207zg_defconfig:
```diff
@@ -0,0 +1,39 @@
+CONFIG_BOARD_NUCLEO_F207ZG=y
+CONFIG_SOC_SERIES_STM32F2X=y
+CONFIG_SOC_STM32F207XG=y
+# 96MHz system clock (highest value to get a precise USB clock)
```
Comment not in line with setting

erwango commented on boards/arm/nucleo_f207zg/nucleo_f207zg.yaml:
```diff
@@ -0,0 +1,7 @@
+arch: arm
+toolchain:
+  - zephyr
+  - gccarmemb
```
Add ram and flash.
Add supported: gpio

erwango commented on arch/arm/soc/st_stm32/stm32f2/Kconfig.defconfig.stm32f207xg:
```diff
@@ -0,0 +1,34 @@
+# Kconfig - ST Microelectronics stm32f207XG MCU
```
Move this commit before board introduction

erwango replied:
Change file name to .stm32f207xx

erwango commented on arch/arm/soc/st_stm32/stm32f2/Kconfig.defconfig.stm32f207xe:
```diff
@@ -0,0 +1,34 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+if SOC_STM32F207XE
```
Add this soc in `arch/arm/soc/st_stm32/stm32f2/Kconfig.defconfig.stm32f207xx`
(cf stm32f417xx soc for instance)

erwango commented on arch/arm/soc/st_stm32/stm32f2/Kconfig.defconfig.stm32f207xe:
```diff
@@ -0,0 +1,34 @@
+# Kconfig - ST Microelectronics stm32f207XE MCU
```
Change commit message to something less personal:
Add STM32F207XE SoC. Used and tested (provide list of tests) on out of tree custom board.

zhaozhongchen commented:
@erwango @ibisz I had update this pr based yours suggestion. please recheck.

ibisz commented:
A branch updated with the uploaded patch over your pr is tested successfully with my custom board(stm32f215zg):

A. Sample 'threads' (gpio, serial):
***** Booting Zephyr OS 1.12.0-rc2 *****
Toggle USR0 LED: Counter = 0
Toggle USR1 LED: Counter = 0
Toggle USR0 LED: Counter = 1
Toggle USR0 LED: Counter = 2
Toggle USR0 LED: Counter = 3
Toggle USR0 LED: Counter = 4
Toggle USR0 LED: Counter = 5
Toggle USR0 LED: Counter = 6
Toggle USR0 LED: Counter = 7
Toggle USR0 LED: Counter = 8
Toggle USR0 LED: Counter = 9
Toggle USR1 LED: Counter = 1
Toggle USR0 LED: Counter = 10
Toggle USR0 LED: Counter = 11
Toggle USR0 LED: Counter = 12
Toggle USR0 LED: Counter = 13
Toggle USR0 LED: Counter = 14
Toggle USR0 LED: Counter = 15
Toggle USR0 LED: Counter = 16
Toggle USR0 LED: Counter = 17
Toggle USR0 LED: Counter = 18
Toggle USR0 LED: Counter = 19
Toggle USR1 LED: Counter = 2
...

B. Sample 'hello_world' (serial):
***** Booting Zephyr OS 1.12.0-rc2 *****
Hello World! arm

@zhaozhongchen, please update your 'stm32f215zg' local branch with the attached patch.

[0001-stm32f2-add-stm32f215xg-soc.zip](https://github.com/zephyrproject-rtos/zephyr/files/2077485/0001-stm32f2-add-stm32f215xg-soc.zip)

erwango commented on drivers/clock_control/Kconfig.stm32:
```diff
@@ -176,6 +176,49 @@ config CLOCK_STM32_PLL2_PREDIV2
+	int "Multiplier factor for PLL VCO output clock"
+	depends on CLOCK_STM32_SYSCLK_SRC_PLL
+	default 192
+	range 192 432
```
According to reference manual range is 64 > 432.

zhaozhongchen replied:
based on RM0033 Page 96, PLLN range is 192 ~ 432.

erwango replied:
So, from RM0033, I can read, page 98:
"VCO output frequency = VCO input frequency × PLLN with 64 ≤ PLLN ≤ 432"

zhaozhongchen replied:
PLLN range is different between RN0033 version 7 and version8.

Version 7: VCO output frequency = VCO input frequency × PLLN with 192 £ PLLN £ 432
Version 8: VCO output frequency = VCO input frequency × PLLN with 64 ≤ PLLN ≤ 432

Beside the PLLN can be select in cubemx is 192 ~ 432.

ibisz replied:
The attachment contains the information regarding the PLLN value range collected from the RM0033 Rev2 December 2010 to
RM0033 Rev8 Mai 2018
and
STM32Cube_FW_F2_V1.1.0 V1.1.1 30-December-2014 to
STM32Cube_FW_F2_V1.7.0 V1.7.0 / 29-September-2017

1. RM0033 REV2 December 2010 to RM0033 REV3 April 2011:
    192 <= PLLN <= 432 with VCO output frequency is between 192 and 432 MHz
2. RM0033 REV4 December 2011 to RM0033 REV6 September 2013:
     64 <= PLLN <= 432 with VCO output frequency is between 64 and 432 MHz
3. RM0033 REV7 February 2015:
   192 <=  PLLN <=  432  with VCO output frequency is between 192 and 432 MHz
4. RM0033 REV8 Mai 2018:
     64 <= PLLN <= 432 with VCO output frequency is between 64 and 432 MHz
5. All FW from STM32Cube_FW_F2_V1.1.0 V1.1.1 30-December-2014 to
         STM32Cube_FW_F2_V1.7.0 V1.7.0 / 29-September-2017
   192 <=  PLLN <=  432  with VCO output frequency is between 192 and 432 MHz
6. STM32Cube
   192 <=  PLLN <=  432

My advice is to keep the 192 to 432 range until a new FW release appears reflecting the latest changes in RM0033 REV8 Mai 2018.

[RCC PLL configuration register (RCC_PLLCFGR) PLLN value range.zip](https://github.com/zephyrproject-rtos/zephyr/files/2108263/RCC.PLL.configuration.register.RCC_PLLCFGR.PLLN.value.range.zip)

erwango replied:
Ok, I confirm that 192 is the good value.
Reference manual will be updated with this value

erwango commented on boards/arm/nucleo_f207zg/doc/nucleo_f207zg.rst:
```diff
@@ -0,0 +1,174 @@
+
+- UART_3_TX : PD8
+- UART_3_RX : PD9
+- PWM_2_CH1 : PA0
```
Remove PWM, I2C and USB for now

erwango reviewed (changes requested):
Some final comments.

erwango reviewed (changes requested):
Last remarks and comments.
Otherwise looks good

ibisz commented on arch/arm/soc/st_stm32/stm32f2/Kconfig.defconfig.series:
```diff
@@ -12,4 +12,14 @@ gsource "arch/arm/soc/st_stm32/stm32f2/Kconfig.defconfig.stm32f2*"
 
+if GPIO_STM32
+
+config GPIO_STM32_PORTD
```
The attached CubeMX project sample is based on the smallest stm32f205rb SOC in stm32f2 series.
In this project, beside the default GPIO A, B, C, the D, H ports are used.
(Pin5 - GPIOH0, Pin6 – GPIOH1, Pin54 – GPIOD2).
Probably this configuration is reflected in this commit.
I can imagine another project/board based on the same SOC with Pin 5,6 and 54 dedicated to OSC_IN, OSC_OUT and UART5_RX respective. In this case configuring the GPIOD and GPIOH is no more necessary.
As I see, at this point we can’t presume the concrete board GPIO configuration.
Moreover, in an extreme situation, some of the default stm32 GPIO A, B, C configuration settings are no more necessary, taking into consideration that the respective pins are occupied by the alternate functions.
[stm32f205rb.zip](https://github.com/zephyrproject-rtos/zephyr/files/2112003/stm32f205rb.zip)

erwango replied:
@ibisz, from what I see in stm32f2**xx.h files, Ports A to I are all supported in all stm32f2 SoCs.
If this is actually the case (tbc) matching Kconfig symbols should be made available/visible to all STM32F2 SoCs (hence in this file).
Technically, this is redundant with what is done in stm32f2.dtsi file (hence should be improved in future), and then both Kconfig symbols and dts files should be in synch.

So as you say, we should not presume what configuration will be required for a board based on one of these SoCs so all symbols should be available (if this is confirmed matching peripherals are actually available on these SoCs)

zhaozhongchen replied:
I select a chip with minimum pin(A, B, C, D, H) in stm32f2 series, So Just leave those gpio port to Kconfig.defconfig.series.

Fewer ports can reduce memory usage although just few bytes. Others SOC can enable other GPIO ports in it's own config.

erwango replied:
@zhaozhongchen , I got your point, but from experience, we prefer to get all GPIO enabled and activated by default at SoC/board level. Then, they can be disabled at application level for optimisation (or at board level for some boards that have really small footprint, 8K of RAM for instance).
There are 2 reasons:
-This is source of error: You don't get any error if you try to use one pin of a deactivated GPIO. So, this is a usually stupid but long issue to debug, specially for a new comer. 
-We prefer to provide stable configurations. If you don't activate it now, someone will activate it later (here at SoC level) because he needs it. Then your configuration will change and you won't be informed. If you need it to be deactivated in you application, enable it here, and deactivate it in your application, you won't get any surprise later.

ibisz commented:
Well done!
@zhaozhongchen , thank you for rebasing/updating the pr and your 'stm32f215zg' local branch with @erwango latest gpio configuration suggestions too.

The 'stm32f215zg' branch is tested successfully with my extra tree custom board:

A. Sample 'hello_world' (serial):
***** Booting Zephyr OS v1.12.0-270-g68d2d5274 *****
Hello World! arm

B. Sample 'threads' (gpio, serial):
***** Booting Zephyr OS v1.12.0-270-g68d2d5274 *****
Toggle USR0 LED: Counter = 0
Toggle USR1 LED: Counter = 0
Toggle USR0 LED: Counter = 1
Toggle USR0 LED: Counter = 2
Toggle USR0 LED: Counter = 3
Toggle USR0 LED: Counter = 4
Toggle USR0 LED: Counter = 5
Toggle USR0 LED: Counter = 6
Toggle USR0 LED: Counter = 7
Toggle USR0 LED: Counter = 8
Toggle USR0 LED: Counter = 9
Toggle USR1 LED: Counter = 1
…

ibisz commented:
@qianfan-Zhao, please rebase.

qianfan-Zhao commented:
rebased

therealprof commented on boards/arm/nucleo_f207zg/support/openocd.cfg:
```diff
@@ -0,0 +1,12 @@
+source [find board/st_nucleo_f2.cfg]
```
This does not exist in OpenOCD.

therealprof commented on boards/arm/nucleo_f207zg/nucleo_f207zg_defconfig:
```diff
@@ -0,0 +1,38 @@
+# however, the board does not have an external oscillator, so just use
+# the 8MHz clock signal coming from integrated STLink
+CONFIG_CLOCK_STM32_HSE_BYPASS=y
+# produce 96MHz clock at PLL output
```
Comment is wrong. I did the math and indeed this will result in the 120MHz mentioned further above.

qianfan-Zhao replied:
Yes it's wrong.  @erwango had report this issue seems I had forget change it.

qianfan-Zhao commented:
@erwang @ibisz 

@therealprof had reported an issue:
> source [find board/st_nucleo_f2.cfg]
> This does not exist in OpenOCD.

I don't have a stm32f2 nucleo board, and I tested this code on my custom board. here is the openocd.cfg:
```
source [find interface/stlink-v2.cfg]
source [find target/stm32f2x.cfg]

$_TARGETNAME configure -event gdb-attach {
	echo "Debugger attaching: halting execution"
	reset halt
	gdb_breakpoint_override hard
}

$_TARGETNAME configure -event gdb-detach {
	echo "Debugger detaching: resuming execution"
	resume
}
```

How to resolve this problem? Seems we should add st_nucleo_f2.cfg to Zephyr's SDK first before this PR merged.

therealprof commented:
I'm using:
```
source [find interface/stlink.cfg]

transport select hla_swd

source [find target/stm32f2x.cfg]

reset_config srst_only

$_TARGETNAME configure -event gdb-attach {
        echo "Debugger attaching: halting execution"
        reset halt
        gdb_breakpoint_override hard
}

$_TARGETNAME configure -event gdb-detach {
        echo "Debugger detaching: resuming execution"
        resume
}
```

qianfan-Zhao commented:
@erwango @therealprof Yes I have a scripts that can works. is there a better way like others nuclo-boards? others has a scripts in Zephyr's SDK.

ibisz commented:
I'm using:

# increase working area to 128KB
set WORKAREASIZE 0x20000

source [find interface/stlink.cfg]

transport select hla_swd

# chip name
set CHIPNAME STM32F215zg

source [find target/stm32f2x.cfg]

$_TARGETNAME configure -event gdb-attach {
        echo "Debugger attaching: halting execution"
        reset halt
        gdb_breakpoint_override hard
}

$_TARGETNAME configure -event gdb-detach {
        echo "Debugger detaching: resuming execution"
        resume
}

therealprof commented:
@erwango I think this will not work:
```source [find target/stm32l4x.cfg]```

OpenOCD is quite picky about finding targets...

therealprof commented:
@qianfan-Zhao The only reason the proposed version does not work is because OpenOCD does not provide a ```board/st_nucleo_f2.cfg``` so what I did was take the contents of ```board/st_nucleo_f1.cfg```, adapted it to the STM32F2 series and used that together with the configuration used by the other Nucleo boards in the Zephyr tree.

qianfan-Zhao commented:
@erwango @ibisz @therealprof please check it

erwango commented:
@therealprof :
> @erwango I think this will not work:
source [find target/stm32l4x.cfg]
OpenOCD is quite picky about finding targets...

Sure, would require an update to f2 (which I guess is present)

@qianfan-Zhao 
> @erwango @therealprof Yes I have a scripts that can works. is there a better way like others nuclo-boards? others has a scripts in Zephyr's SDK.

Zephyr SDK is using OpenOCD master branch. We don't intend now to go in branch.
So we have to use whatever is in openocd ...

ibisz commented:
Tested with my extra tree custom board (stm32f215zg):
After applying the ‘stm32f2-add-stm32f215xg-soc.patch’ over the stm32f2 branch and 
copying the openocd.cfg from the nucleo_f207zg board to my custom board, the hello_world and threads samples passes the test successfully:
```
***** Booting Zephyr OS v1.12.0-441-g0c2cb6cdd *****
Hello World! arm
```
and
```
***** Booting Zephyr OS v1.12.0-441-g0c2cb6cdd *****
Toggle USR0 LED: Counter = 0
Toggle USR1 LED: Counter = 0
Toggle USR0 LED: Counter = 1
Toggle USR0 LED: Counter = 2
Toggle USR0 LED: Counter = 3
Toggle USR0 LED: Counter = 4
Toggle USR0 LED: Counter = 5
Toggle USR0 LED: Counter = 6
Toggle USR0 LED: Counter = 7
Toggle USR0 LED: Counter = 8
Toggle USR0 LED: Counter = 9
Toggle USR1 LED: Counter = 1
Toggle USR0 LED: Counter = 10
Toggle USR0 LED: Counter = 11
```
The OpenOCD log contains a WARNING related to the interface/stlink-v2-1.cfg, see below:
```
Ninja flash
[99/105] Linking C executable zephyr\zephyr_prebuilt.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       15808 B         1 MB      1.51%
            SRAM:        7812 B       128 KB      5.96%
        IDT_LIST:         216 B         2 KB     10.55%
[104/105] Flashing stm32f215zg_custom
Using runner: openocd
GNU MCU Eclipse 64-bits Open On-Chip Debugger 0.10.0+dev-00487-gaf359c18 (2018-05-12-19:30)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
WARNING: interface/stlink-v2-1.cfg is deprecated, please switch to interface/stlink.cfg
Info : auto-selecting first available session transport "hla_swd". To override use 'transport select <transport>'.
adapter speed: 1000 kHz
adapter_nsrst_delay: 100
Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
none separate
Info : Unable to match requested speed 1000 kHz, using 950 kHz
Info : Unable to match requested speed 1000 kHz, using 950 kHz
Info : clock speed 950 kHz
Info : STLINK v2 JTAG v28 API v2 SWIM v7 VID 0x0483 PID 0x3748
Info : using stlink api v2
Info : Target voltage: 3.215213
Info : stm32f2x.cpu: hardware has 6 breakpoints, 4 watchpoints
Info : Listening on port 3333 for gdb connections
    TargetName         Type       Endian TapName            State
--  ------------------ ---------- ------ ------------------ ------------
 0* stm32f2x.cpu       hla_target little stm32f2x.cpu       running
target halted due to debug-request, current mode: Thread
xPSR: 0x01000000 pc: 0x080014a8 msp: 0x200005f0
auto erase enabled
Info : device id = 0x20036411
Info : flash size = 1024kbytes
target halted due to breakpoint, current mode: Thread
xPSR: 0x61000000 pc: 0x20000044 msp: 0x200005f0
wrote 16384 bytes from file …/stm32f215zg_custom/threads/zephyr/zephyr.bin in 0.815919s (19.610 KiB/s)
target halted due to debug-request, current mode: Thread
xPSR: 0x01000000 pc: 0x08001a00 msp: 0x20001440
target halted due to breakpoint, current mode: Thread
xPSR: 0x61000000 pc: 0x2000002e msp: 0x20001440
verified 15808 bytes in 0.136895s (112.769 KiB/s)
shutdown command invoked
```

therealprof commented:
I haven't been able to test the new `openocd.cfg` yet but the code works fine on my Nucleo-F207ZG and I even have some more code in my tree to enable more functionality including networking which works fine after working around #8668. Looking for ward to getting this merged.

therealprof reviewed (approved):
Works fine for me on a Nucleo-F207ZG.

ibisz reviewed (approved):
Works error free on a extra tree stm32f215zg based custom board.
